lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp

```diff
@@ -687,6 +693,64 @@ void InitLtcModuleBindings(py::module m) {
   m.def("_ltc_enable_thread_pool", []() {
     FLAGS_torch_lazy_use_thread_pool = true;
   });
+  /*
+   * Return tensor ids and tensors for DeviceData nodes.
+   */
+  m.def("_get_ltc_tensors_ts_device_data_node",
+        [](const std::vector<at::Tensor>& tensors) {
+          std::vector<Node*> roots;
+          for (auto& tensor : tensors) {
+            auto xtensor = TryGetLtcTensor(tensor);
+            roots.push_back(xtensor.GetIrValue().node.get());
+          }
+          auto post_order = Util::ComputePostOrder(roots);
+          std::vector<int64_t> tensor_ids;
+          std::vector<at::IValue> ivalues;
+          for (auto nodeptr : post_order) {
+            if (nodeptr->op() == *torch::lazy::ltc_device_data) {
+              const auto* device_data_node = torch::lazy::NodeCast<torch::lazy::DeviceData>(nodeptr, *torch::lazy::ltc_device_data);
+              auto infoptr = device_data_node->data()->info();
+              auto deviceDataInfoPtr = (torch::lazy::LazyGraphExecutor::DeviceDataInfo*) infoptr;
+              tensor_ids.push_back(deviceDataInfoPtr->tensor_id);
+
+              auto* tsDataPtr = (torch_lazy_tensors::compiler::TSData*) device_data_node->data().get();
+              if (tsDataPtr->HasValue()) {
+                ivalues.push_back(tsDataPtr->data());
+              } else {
+                CHECK(tsDataPtr->scalar.has_value());
+                ivalues.push_back(tsDataPtr->scalar.value());
+              }
+            }
+          }
+          return std::make_pair(tensor_ids, ivalues);
+        });
+  m.def("_get_graph_hash", [](const std::vector<at::Tensor>& tensors) {
+    std::vector<LazyTensor> xtensors;
+    for (auto& tensor : tensors) {
+      xtensors.push_back(TryGetLtcTensor(tensor));
+    }
+    auto hash = LazyGraphExecutor::Get()->GetGraphHash(xtensors);
+    std::string bin((const char*) &hash, sizeof(hash));
+    return py::bytes(bin);
+  });
+  m.def("_run_cached_graph", [](const std::string& hash_str, const std::vector<at::IValue>& graph_inputs) {
```

Comment:
oh, this is also torchscript specific. So we should think about if we plan to extend to also support XLA.  It might be good to show this to @JackCaoG first and see if it's clear how to do it, we could put a TODO here or else mention that we don't have plans to support it yet but it could be future work?

Reply:
I'll add a TODO. We should integrate with XLA when the backend is ready

Reply:
they probably have something very similar though their inputs might just always be tensors and then unwrap them into scalars in their execution engine.